### ISSUE TITLE
Code that the termination checker only accepts if you engage in a bunch of duplication

### ISSUE THREAD

**Issue by Soares** [agda/agda#4511]
Consider the following module. It has the following properties:

1. The termination checker rejects it
2. There exists a modified program that does the same thing which the termination checker accepts
3. Producing the modified program requires duplicating code.

In this particular module, the types are trivial and the problem may look asinine. However, it's been (heavily) distilled from a more general application, where I'm bumping into this issue a handful of times, and the necessary code duplication to work around it is quite substantial.

```agda
module Main where

record Null : Set where
  constructor null

record A : Set
record B (a₁ a₂ : A) : Set
terminal : A

record A where
  coinductive
  field a : A
  field b : B a terminal
open A

record B a₁ a₂ where
  field n : Null
open B

any : ∀ {a : A} -> B a terminal
any .n = null

terminal .a = terminal
terminal .b = any {terminal}
```

In particular:

- `A` and `B` are defined using induction-induction. Also, we've singled out a special `terminal` `A`
- To get an `A` you must show (coinductively) some `a` such that `B a terminal`
- `any` shows how to get a `B a terminal` for any `a : A`
- The termination checker objects if we try to define `special` in terms of `any {terminate}`

If `any` can be implemented at all, then `terminal` can of course be re-implemented as a special case of `any`. (More explicitly, change the last line to `terminal .b .n = null`, and the program succeeds). However, if `any` is a complicated definition, the code duplication here may be substantial.

Currently I don't understand the termination checker well enough to know whether this issue is a bug or a feature request, or perhaps just an issue of me misusing the language. FWIW, I'd also love a pointer in the direction of any relevant docs; I wasn't able to glean anything useful from [the termination checker docs](https://agda.readthedocs.io/en/v2.6.0.1/language/termination-checking.html) but perhaps I'm being daft or neglecting some other resource.

**Comment by gallais** [comment]
Looks like https://github.com/agda/agda/issues/4391 to me. Can you try to see whether your problem persists with the current dev version on the master branch?

**Comment by Soares** [comment]
I'll go check. In the interim, though, please note that I can exhibit this problem without using guardedness or coinduction. (The types get even more asinine, but the termination checker still chokes.) Behold:

```agda
-- AFAICT the guardedness setting has no effect on this code
module Main where

record Null : Set where
  constructor null

record A : Set
record B (a₁ a₂ : A) : Set
terminal : A

record A where
  field b : B terminal terminal
open A

record B a₁ a₂ where
  field n : Null
open B

any : ∀ {a : A} -> B a terminal
any .n = null

terminal .b = any {terminal}
```

**Comment by Soares** [comment]
(The Agda dev version has been building on my machine for the last 2h, with no sign of stopping, and I will shortly have to hit the road. I'm not sure when I'll be able to get it installed & test the above code. If anyone else can check in the interim, I'd be much obliged.)

**Comment by Soares** [comment]
I have managed to confirm that this bug is *not* fixed on `Agda version 2.6.1-fed2ee1`.

**Comment by andreasabel** [comment]
> terminal .b = any {terminal}

The termination checker rejects this because it fears the worst:  That `any` makes a call to `terminal`...

Here, `terminal` constructs a `record` subject to eta equality, thus, the definition is equivalent to:
```agda
terminal = record { b = any {terminal} }
```
Now, it is not hard to see how you can construct an infinite reduction sequence from this (replace `terminal` by its definition in its definition and notice that you can do this again, and again, and again...).

I am closing this issue now, but any _concrete_ proposal how to change the documentation is very welcome, ideally in a PR!
